lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/address/ContainerBasedLocalAddressProvider.java

```diff
@@ -0,0 +1,114 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.reef.wake.remote.address;
+
+import org.apache.reef.tang.Configuration;
+import org.apache.reef.tang.Tang;
+
+import javax.inject.Inject;
+import java.io.File;
+import java.io.IOException;
+import java.net.Inet4Address;
+import java.net.UnknownHostException;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * A LocalAddressProvider that uses <code>Inet4Address.getLocalHost().getHostAddress()</code>.
+ */
+public final class ContainerBasedLocalAddressProvider implements LocalAddressProvider {
+
+  private static final Logger LOG = Logger.getLogger(ContainerBasedLocalAddressProvider.class.getName());
+
+  private String cached = null;
+
+  /**
+   * The constructor is for Tang only.
+   */
+  @Inject
+  private ContainerBasedLocalAddressProvider() {
+    LOG.log(Level.FINE, "Instantiating ContainerBasedLocalAddressProvider");
+  }
+
+  @Override
+  public synchronized String getLocalAddress() {
+    if (null == cached) {
+      String ipAddressPath = System.getenv("HOST_IP_ADDR_PATH");
+      LOG.log(Level.INFO, "IpAddressPath is {0}", ipAddressPath);
+      if (ipAddressPath != null) {
+        File ipAddressFile = new File(ipAddressPath);
+        if (ipAddressFile.exists() && ipAddressFile.isFile()) {
+          String filePath = expandEnvironmentVariables(ipAddressPath);
+          try {
+            cached = readFile(filePath, StandardCharsets.UTF_8);
+          } catch (IOException e) {
+            String message = String.format("Exception when attempting to read file %s", filePath);
+            LOG.log(Level.SEVERE, message, e);
+            throw new RuntimeException(message);
+          }
+        } else {
+          final String message = "HOST_IP_ADDR_PATH points to invalid path.";
+          LOG.log(Level.SEVERE, message);
+          throw new RuntimeException(message);
+        }
+      } else {
+        final String message = "Environment variable must be set for HOST_IP_ADDR_PATH";
+        LOG.log(Level.SEVERE, message);
+        throw new RuntimeException(message);
+      }
+    }
+    assert null != cached;
+    LOG.log(Level.INFO, "Returning ContainerBasedLocalAddressProvider.getLocalAddress() as " + cached);
+    return cached;
+  }
+
+  @Override
+  public Configuration getConfiguration() {
+    return Tang.Factory.getTang().newConfigurationBuilder()
+        .bind(LocalAddressProvider.class, ContainerBasedLocalAddressProvider.class)
+        .build();
+  }
+
+  @Override
+  public String toString() {
+    return "ContainerBasedLocalAddressProvider:" + this.getLocalAddress();
+  }
+
+  public static String expandEnvironmentVariables(String text) {
```

Comment:
why is this method here?

Reply:
Because it is used above. In windows, the environment variable %VAR% is not being expanded. So I have to do it here.

---
In reply to: [196586135](https://github.com/sharathmalladi/reef/pull/5#discussion_r196586135) [](ancestors = 196586135)

Reply:
in that case, it could be private function. no need to be static as well.

---
In reply to: [196967727](https://github.com/sharathmalladi/reef/pull/5#discussion_r196967727) [](ancestors = 196967727,196586135)

Reply:
Change from public static to private

---
In reply to: [196971191](https://github.com/sharathmalladi/reef/pull/5#discussion_r196971191) [](ancestors = 196971191,196967727,196586135)